# Modeling

I transformed the molecules using Color Refinement.

[Here is a demo of Color Refinement](https://github.com/parkerburchett/TDC-DeepLearning/blob/main/utils/ColorRefinement_tutorial.ipynb)

I then tested some simple MLP with different loss and optimizer functions.

I then transformed the bags of colors for each neighborhood (0-10) using PCA and trained a "out of the box" LGBM regressor model. That model had a AUPRC of .8375+- .0093. That is a ~.09 improvement on the model that TDC is currently using.

These are the outcomes TDC's current model.
https://tdcommons.ai/benchmark/admet_group/10cyp2c9i/


I ran this in Google Colab and then saved it to this repo.

In [1]:
!pip install pysmiles
!pip install pyTDC
!git clone https://github.com/parkerburchett/TDC-DeepLearning

fatal: destination path 'TDC-DeepLearning' already exists and is not an empty directory.


In [2]:
from google.colab import drive
drive.mount('/content/drive')
import networkx as nx
import tensorflow as tf
from sklearn.metrics import precision_recall_curve, auc
from tensorflow.keras.optimizers import SGD
import pandas as pd
import numpy as np
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split
import pickle
from pysmiles import read_smiles
from tdc.single_pred import ADME
import os
os.chdir('/content/TDC-DeepLearning/')
from utils import ColorRefinement, utils

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
color_graphs = pickle.load(open('/content/drive/MyDrive/SpringBoard/Therapeutic Data Commons Projects/data/color_graphs.pkl', 'rb'))

# This is the 0-10 hop dfs of each of the molecules. Since it takes 45 minutes to run I ran it once and am just loading it each time

In [4]:
color_graphs[0] # a single molecule

0    1    2    3    4    5     ... 4994 4995 4996 4997 4998 4999
iteration_number                                ...                              
color_0             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_1             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_2             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_3             0    0    0    0    0    0  ...    0    0    0    0    1    0
color_4             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_5             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_6             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_7             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_8             0    0    0    0    0    0  ...    0    0    0    0    0    0
color_9             0    0    0    0    0    0  ...    0    0    0    0    0    0

[10 rows x 5000 columns]

All of the feautre vectors are very spare, typically 95% + zeros.

In [5]:
def get_n_dem_embedding(color_graphs, n):
  embeddings = []
  for g in color_graphs:
    embeddings.append(g.values[n,:])
  return pd.DataFrame(np.array(embeddings))


In [6]:
feature_dfs = [get_n_dem_embedding(color_graphs, n) for n in range(0,10)]

In [7]:
feature_dfs[0]

0    1    2    3    4    5    6     ... 4993 4994 4995 4996 4997 4998 4999
0        0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
1        0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
2        0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
3        0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
4        0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...
12660    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
12661    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
12662    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
12663    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
12664    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0

[12665 rows x 5000 columns]

Each row is a molecule, and each colum is a count of the final colors of each of it's nodes.

In [13]:
feature_dfs[0].sum(axis=0).sort_values().tail(10)

3327     8731.0
3398     9516.0
2357    11544.0
4882    12734.0
3448    18956.0
4073    25399.0
3321    28378.0
3723    41797.0
884     50259.0
2659    78256.0
dtype: float64

The O hop assigned each unique kind of node, (unique element, h_count, aromatic, and charge) -> a unique color. There are no false negatives, there is a 1/5000 chance of a false positive.
 
Otherwise, if you see the number 2659 assigned to different nodes in two molecules you can be very confident that those nodes are identical


In [14]:
feature_dfs[3].sum(axis=0).sort_values().tail(10)

4955     945.0
4258    1075.0
42      1080.0
4769    1096.0
1058    1386.0
285     1390.0
81      1406.0
2775    1441.0
450     2186.0
982     3922.0
dtype: float64

The 3 refers to the 3 hop neighborhoods, this means that when you see two nodes having the color 982 there is a 5000-(3+1)/5000 chance that the local 3 hop neighborhood around those nodes is identical.
 
I am using this method to convert arbitrarily sized local neighborhoods into features.
 
The 3 hop neighborhood converts each 3 hop neighborhood into a color and then a graph is encoded as a count of each of the [0, 4999] possible colors.
 


In [15]:
# for index, df in enumerate(feature_dfs):
#   df.to_csv(f'/content/drive/MyDrive/SpringBoard/Therapeutic Data Commons Projects/data/feature_hop_{index}.csv')
#   print('saved a df')

In [8]:
df = pd.read_csv('/content/TDC-DeepLearning/data/cleaned_raw_data.csv')
df

SMILE  Target
0                 CCCC(=O)Nc1ccc(N2CCN(CC)CC2)c(Cl)c1.Cl       0
1             CC(=O)N(c1ccc2oc(=O)sc2c1)S(=O)(=O)c1cccs1       1
2                       Clc1ccccc1-c1nc(-c2ccccc2)n[nH]1       1
3        COc1ccccc1C(c1nnnn1C(C)(C)C)N1CCN(Cc2ccncc2)CC1       1
4                   COc1ccccc1CNC(=O)Cn1nnc(-c2ccncc2)n1       1
...                                                  ...     ...
12660  CCN1C(=O)[C@@H]2[C@@H](CC[C@@H]3C(=O)C=C[C@@H]...       0
12661  O=C1[C@H]2CC=C3[C@@H]([C@H](O)[C@H]4O[C@@H]4C3...       0
12662  CCN1C(=O)[C@H]2CC=C3[C@@H]([C@H](O)[C@H]4O[C@@...       1
12663  C[C@H](c1ccccc1)N1C(=O)[C@@H]2[C@@H](CC[C@@H]3...       0
12664                        COC(=O)NC/C=C\c1nc(CCCO)co1       0

[12665 rows x 2 columns]

In [9]:
features = pd.concat(feature_dfs[3:], axis=1).astype(np.int8)
# using the colors of hop 3 to 10
print(features.dtypes[0])
print(features.shape)


0    int8
0    int8
0    int8
0    int8
0    int8
0    int8
0    int8
dtype: object
(12665, 35000)


Each molecule is converted into a vector of the 3 hop neighborhoods appended to the 4 hop neighborhoods ... 10 hop neighborhood. That vector is then passed into the MLP.

In [22]:
from tensorflow.keras.layers import *

def compute_auprc(model, X_test, y_test):
    # https://stats.stackexchange.com/questions/157012/area-under-precision-recall-curve-auc-of-pr-curve-and-average-precision-ap
    precision, recall, thresholds = precision_recall_curve(y_test, model.predict(X_test))
    area = auc(recall, precision)
    return area

def get_optimizer():
  # https://towardsdatascience.com/simple-guide-to-hyperparameter-tuning-in-neural-networks-3fe03dad8594
	learning_rate = 0.1
	decay_rate = learning_rate / 300
	momentum = 0.8
	sgd = SGD(lr=learning_rate, momentum=momentum, decay=decay_rate, nesterov=False)
	return sgd

def create_baseline():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))
  model.compile(loss='binary_crossentropy', optimizer=get_optimizer(), metrics=['accuracy'])
  return model


def create_relu():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='relu'))
  model.compile(loss='binary_crossentropy', optimizer=get_optimizer(), metrics=['accuracy'])
  return model

def create_longer_model():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))
  model.compile(loss='binary_crossentropy', optimizer=get_optimizer(), metrics=['accuracy'])
  return model


def create_longer_thin_model():
  model = Sequential()
  model.add(Dense(32, activation='relu'))
  model.add(Dense(32, activation='relu'))
  model.add(Dense(32, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))
  model.compile(loss='binary_crossentropy', optimizer=get_optimizer(), metrics=['accuracy'])
  return model

def create_soft_max():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='softmax'))
  model.compile(loss='binary_crossentropy', optimizer=get_optimizer(), metrics=['accuracy'])
  return model

def create_adam_optimzer():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='softmax'))
  model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
  return model

def create_poisson_loss():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))
  model.compile(loss='poisson', optimizer=get_optimizer(), metrics=['accuracy'])
  return model

def create_kl_divergence():
  model = Sequential()
  model.add(Dense(64, activation='relu'))
  model.add(Dense(64, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))
  model.compile(loss='kl_divergence', optimizer=get_optimizer(), metrics=['accuracy'])
  return model




In [62]:
def evaluate(model_function, X, y):
  scores = []
  for i in range(5):
    model = model_function()
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=i)
    for s in range(10):
      model = model_function()
      model.fit(X_train, y_train, validation_data=(X_test, y_test))
      score = compute_auprc(model, X_test, y_test)
      print(f'AUPRC : {round(score, 4)}')
      scores.append(score)
      del model
  
  print(f'Average AUPRC: {round(np.average(scores), 4)} +- {round(np.std(scores), 4)} ')



In [20]:
X, y = features.values, df['Target'].values

In [63]:
evaluate(create_baseline, X, y)

317/317 [==============================] - 1s 4ms/step - loss: 0.5597 - accuracy: 0.7033 - val_loss: 0.5034 - val_accuracy: 0.7604
AUPRC : 0.8062
317/317 [==============================] - 1s 3ms/step - loss: 0.5581 - accuracy: 0.7054 - val_loss: 0.4930 - val_accuracy: 0.7694
AUPRC : 0.8153
317/317 [==============================] - 1s 3ms/step - loss: 0.5606 - accuracy: 0.7054 - val_loss: 0.4947 - val_accuracy: 0.7671
AUPRC : 0.8188
317/317 [==============================] - 2s 5ms/step - loss: 0.5606 - accuracy: 0.7084 - val_loss: 0.5058 - val_accuracy: 0.7600
AUPRC : 0.816
317/317 [==============================] - 1s 4ms/step - loss: 0.5586 - accuracy: 0.7015 - val_loss: 0.4944 - val_accuracy: 0.7683
AUPRC : 0.8192
317/317 [==============================] - 1s 4ms/step - loss: 0.5579 - accuracy: 0.7062 - val_loss: 0.4919 - val_accuracy: 0.7627
AUPRC : 0.8139
317/317 [==============================] - 1s 3ms/step - loss: 0.5549 - accuracy: 0.7052 - val_loss: 0.5037 - val_accuracy: 0

In [64]:
evaluate(create_relu, X, y)

317/317 [==============================] - 1s 3ms/step - loss: 8.2771 - accuracy: 0.4563 - val_loss: 8.0250 - val_accuracy: 0.4737
AUPRC : 0.5341
317/317 [==============================] - 1s 4ms/step - loss: 8.2671 - accuracy: 0.4571 - val_loss: 8.0250 - val_accuracy: 0.4737
AUPRC : 0.5786
317/317 [==============================] - 1s 4ms/step - loss: 8.2822 - accuracy: 0.4563 - val_loss: 8.0250 - val_accuracy: 0.4737
AUPRC : 0.5105
317/317 [==============================] - 1s 4ms/step - loss: 8.2803 - accuracy: 0.4561 - val_loss: 8.0250 - val_accuracy: 0.4737
AUPRC : 0.525
317/317 [==============================] - 1s 4ms/step - loss: 8.2731 - accuracy: 0.4567 - val_loss: 8.0250 - val_accuracy: 0.4737
AUPRC : 0.5446
317/317 [==============================] - 1s 4ms/step - loss: 8.2755 - accuracy: 0.4563 - val_loss: 8.0250 - val_accuracy: 0.4737
AUPRC : 0.527
317/317 [==============================] - 2s 4ms/step - loss: 8.2795 - accuracy: 0.4563 - val_loss: 8.0250 - val_accuracy: 0.

In [65]:
evaluate(create_longer_model, X, y)

317/317 [==============================] - 2s 4ms/step - loss: 0.5748 - accuracy: 0.6897 - val_loss: 0.5124 - val_accuracy: 0.7521
AUPRC : 0.8003
317/317 [==============================] - 2s 4ms/step - loss: 0.5762 - accuracy: 0.6861 - val_loss: 0.5352 - val_accuracy: 0.7505
AUPRC : 0.804
317/317 [==============================] - 2s 4ms/step - loss: 0.5752 - accuracy: 0.6915 - val_loss: 0.5094 - val_accuracy: 0.7643
AUPRC : 0.8011
317/317 [==============================] - 2s 4ms/step - loss: 0.5760 - accuracy: 0.6873 - val_loss: 0.5035 - val_accuracy: 0.7509
AUPRC : 0.8108
317/317 [==============================] - 2s 4ms/step - loss: 0.5786 - accuracy: 0.6847 - val_loss: 0.5095 - val_accuracy: 0.7513
AUPRC : 0.8089
317/317 [==============================] - 2s 4ms/step - loss: 0.5788 - accuracy: 0.6836 - val_loss: 0.5192 - val_accuracy: 0.7481
AUPRC : 0.8139
317/317 [==============================] - 2s 4ms/step - loss: 0.5749 - accuracy: 0.6885 - val_loss: 0.5135 - val_accuracy: 0

In [66]:
evaluate(create_longer_thin_model,X, y)

317/317 [==============================] - 2s 4ms/step - loss: 0.5842 - accuracy: 0.6737 - val_loss: 0.5582 - val_accuracy: 0.7177
AUPRC : 0.8016
317/317 [==============================] - 2s 4ms/step - loss: 0.5757 - accuracy: 0.6873 - val_loss: 0.5038 - val_accuracy: 0.7631
AUPRC : 0.811
317/317 [==============================] - 2s 4ms/step - loss: 0.5849 - accuracy: 0.6819 - val_loss: 0.5057 - val_accuracy: 0.7663
AUPRC : 0.8058
317/317 [==============================] - 2s 4ms/step - loss: 0.5830 - accuracy: 0.6841 - val_loss: 0.5171 - val_accuracy: 0.7485
AUPRC : 0.8025
317/317 [==============================] - 2s 4ms/step - loss: 0.5754 - accuracy: 0.6916 - val_loss: 0.5049 - val_accuracy: 0.7584
AUPRC : 0.8037
317/317 [==============================] - 2s 4ms/step - loss: 0.5865 - accuracy: 0.6777 - val_loss: 0.5124 - val_accuracy: 0.7481
AUPRC : 0.7917
317/317 [==============================] - 2s 4ms/step - loss: 0.5771 - accuracy: 0.6878 - val_loss: 0.5183 - val_accuracy: 0

In [67]:
evaluate(create_soft_max, X, y)

317/317 [==============================] - 1s 3ms/step - loss: 0.5605 - accuracy: 0.4559 - val_loss: 0.4883 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5590 - accuracy: 0.4559 - val_loss: 0.5007 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 2s 4ms/step - loss: 0.5613 - accuracy: 0.4559 - val_loss: 0.4945 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 4ms/step - loss: 0.5617 - accuracy: 0.4559 - val_loss: 0.4951 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5588 - accuracy: 0.4559 - val_loss: 0.4942 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 4ms/step - loss: 0.5632 - accuracy: 0.4559 - val_loss: 0.5070 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5617 - accuracy: 0.4559 - val_loss: 0.4944 - val_accuracy: 

In [68]:
evaluate(create_adam_optimzer, X, y)

317/317 [==============================] - 2s 4ms/step - loss: 0.5684 - accuracy: 0.4559 - val_loss: 0.5157 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5666 - accuracy: 0.4559 - val_loss: 0.5076 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5686 - accuracy: 0.4559 - val_loss: 0.5073 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5719 - accuracy: 0.4559 - val_loss: 0.5095 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 2s 4ms/step - loss: 0.5701 - accuracy: 0.4559 - val_loss: 0.5040 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5660 - accuracy: 0.4559 - val_loss: 0.5073 - val_accuracy: 0.4737
AUPRC : 0.7369
317/317 [==============================] - 1s 3ms/step - loss: 0.5683 - accuracy: 0.4559 - val_loss: 0.5111 - val_accuracy: 

In [69]:
evaluate(create_poisson_loss, X, y)

317/317 [==============================] - 1s 3ms/step - loss: 0.7469 - accuracy: 0.6951 - val_loss: 0.7222 - val_accuracy: 0.7659
AUPRC : 0.8097
317/317 [==============================] - 1s 3ms/step - loss: 0.7470 - accuracy: 0.6934 - val_loss: 0.7180 - val_accuracy: 0.7635
AUPRC : 0.8031
317/317 [==============================] - 1s 3ms/step - loss: 0.7432 - accuracy: 0.6983 - val_loss: 0.7261 - val_accuracy: 0.7615
AUPRC : 0.8138
317/317 [==============================] - 1s 3ms/step - loss: 0.7444 - accuracy: 0.6959 - val_loss: 0.7310 - val_accuracy: 0.7584
AUPRC : 0.8117
317/317 [==============================] - 1s 3ms/step - loss: 0.7443 - accuracy: 0.7046 - val_loss: 0.7299 - val_accuracy: 0.7580
AUPRC : 0.7954
317/317 [==============================] - 1s 3ms/step - loss: 0.7448 - accuracy: 0.6955 - val_loss: 0.7256 - val_accuracy: 0.7596
AUPRC : 0.798
317/317 [==============================] - 1s 3ms/step - loss: 0.7455 - accuracy: 0.7014 - val_loss: 0.7241 - val_accuracy: 0

In [70]:
evaluate(create_kl_divergence, X, y)

317/317 [==============================] - 1s 3ms/step - loss: 0.0082 - accuracy: 0.4561 - val_loss: 2.0764e-04 - val_accuracy: 0.4737
AUPRC : 0.5543
317/317 [==============================] - 1s 3ms/step - loss: 0.0097 - accuracy: 0.4567 - val_loss: 2.2172e-04 - val_accuracy: 0.4737
AUPRC : 0.5591
317/317 [==============================] - 1s 3ms/step - loss: 0.0094 - accuracy: 0.4557 - val_loss: 2.2342e-04 - val_accuracy: 0.4737
AUPRC : 0.5608
317/317 [==============================] - 1s 3ms/step - loss: 0.0088 - accuracy: 0.4562 - val_loss: 2.0361e-04 - val_accuracy: 0.4737
AUPRC : 0.5626
317/317 [==============================] - 1s 3ms/step - loss: 0.0094 - accuracy: 0.4553 - val_loss: 2.1510e-04 - val_accuracy: 0.4737
AUPRC : 0.551
317/317 [==============================] - 2s 5ms/step - loss: 0.0109 - accuracy: 0.4561 - val_loss: 2.5403e-04 - val_accuracy: 0.4737
AUPRC : 0.5497
317/317 [==============================] - 1s 3ms/step - loss: 0.0096 - accuracy: 0.4558 - val_loss: 

# Modeling PCA Features

In [80]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# pca_dfs = [pd.DataFrame(PCA(500).fit_transform(df)) for df in feature_dfs]
# # note: this violates cross validation rules. the Transformer is trained on the whole data

## I need to use PCA becasuse otherwise the modeling would take much to long.

In [81]:
# pca_features = pd.concat(pca_dfs, axis=1)

In [82]:
# pca_features

0         1         2    ...       497       498       499
0     -8.263492  0.963605 -0.888452  ...  0.086430  0.124055  0.130666
1      0.898335 -2.586220  2.007668  ...  0.061278  0.076817  0.134793
2      4.493463 -2.558734 -2.572918  ...  0.131846 -0.021940 -0.019103
3      0.378278  2.553361 -0.331753  ... -0.109355  0.102852 -0.108537
4      1.681549 -0.250663 -0.265615  ... -0.174192 -0.003170 -0.230648
...         ...       ...       ...  ...       ...       ...       ...
12660 -7.235418 -2.289271  1.249360  ...  0.024352  0.035861  0.002760
12661 -3.474046  1.872275  2.617426  ...  0.048276 -0.117995 -0.133305
12662 -8.063706 -0.096515  2.566241  ...  0.011053  0.004408  0.096661
12663 -2.769751 -0.268407  1.856931  ...  0.054479  0.089447 -0.115508
12664 -5.189814 -1.747349 -1.401867  ... -0.004868  0.006856  0.072575

[12665 rows x 5000 columns]

In [107]:
from lightgbm import LGBMRegressor
def evaluate_lbg(model_function, X, y):
  scores = []
  for i in range(10): # smaller sample since it takes much longer
    model = LGBMRegressor(silent=False)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=i)
    model.fit(X_train, y_train)
    score = compute_auprc(model, X_test, y_test)
    print(f'AUPRC : {round(score, 4)}')
    scores.append(score)
    del model

  print(f'Average AUPRC: {round(np.average(scores), 4)} +- {round(np.std(scores), 4)} ')

In [108]:
%%time
evaluate_lbg(LGBMRegressor, pca_features.values, y) # looks very promising

AUPRC : 0.8589
AUPRC : 0.8289
AUPRC : 0.8433
AUPRC : 0.8282
AUPRC : 0.8376
AUPRC : 0.8405
AUPRC : 0.8307
AUPRC : 0.8266
AUPRC : 0.8402
AUPRC : 0.8517
Average AUPRC: 0.8387 +- 0.0101 
CPU times: user 30min 17s, sys: 7.5 s, total: 30min 25s
Wall time: 15min 39s


# PCA not violating cross validation

In [136]:
features = features.astype(np.int16)

In [19]:
from sklearn.decomposition import PCA

y = df['Target'].values

def get_pca_transformed_split_data(features, feature_df, seed):

  np.random.seed(seed)
  test_indices = np.random.choice(features.index, int(features.shape[0]*.2), replace=False) # set aside a 20% test set.
  test_index_mask = np.array([i in test_indices for i in features.index])
  train_index_mask = ~test_index_mask

  transformed_train_features = []
  transformed_test_features = []
  for df in feature_dfs:
    pca_model = PCA(500).fit(df[train_index_mask])
    print('fit pca model')

    transformed_train_df = pd.DataFrame(pca_model.transform(df[train_index_mask]))
    transformed_test_df = pd.DataFrame(pca_model.transform(df[test_index_mask]))

    transformed_train_features.append(transformed_train_df)
    transformed_test_features.append(transformed_test_df)


  X_train = pd.concat(transformed_train_features, axis=1).values
  y_train = y[train_index_mask]


  X_test = pd.concat(transformed_test_features, axis=1).values
  y_test = y[test_index_mask]

  return X_train, X_test, y_train, y_test


X_train, X_test, y_train, y_test = get_pca_transformed_split_data(features, feature_dfs, 55)


fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model


In [20]:
X_train.shape, y_train.shape

((10132, 5000), (10132,))

In [21]:
from lightgbm import LGBMRegressor
model = LGBMRegressor()
model.fit(X_train, y_train)
print('fit lgb model')



fit lgb model


NameError: ignored

In [23]:
score = compute_auprc(model, X_test, y_test)
print(f'AUPRC : {round(score, 4)}')

AUPRC : 0.8531


In [25]:
%%time
scores = []
for i in range(10):
  X_train, X_test, y_train, y_test = get_pca_transformed_split_data(features, feature_dfs, i)
  model = LGBMRegressor()
  model.fit(X_train, y_train)
  print('fit lgb model')
  score = compute_auprc(model, X_test, y_test)
  print(f'AUPRC : {round(score, 4)}')
  scores.append(score)

print(f'Average AUPRC: {round(np.average(scores), 4)} +- {round(np.std(scores), 4)} ')

fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit lgb model
AUPRC : 0.8499
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit lgb model
AUPRC : 0.8307
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit lgb model
AUPRC : 0.8408
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit lgb model
AUPRC : 0.8433
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit lgb model
AUPRC : 0.8458
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit pca model
fit lgb model
A

# current best model


ColorRefinement (hops 0-10) -> PCA (500 features) -> 5000 long PCA transformed features of the color refinements -> out of the box LGBM 


Average AUPRC: 0.8375 +- 0.0093.

A significant improvement on the current https://tdcommons.ai/benchmark/admet_group/10cyp2c9i/

CNN (DeepPurpose)	Kexin Huang	GitHub, Paper	226,625	0.713 ± 0.006
